select witnesses only for unique inps

paweljakubas opened the pull request:
<!--
Detail in a few bullet points the work accomplished in this PR.

Before you submit, don't forget to:

* Make sure the GitHub PR fields are correct:
   ✓ Set a good Title for your PR.
   ✓ Assign yourself to the PR.
   ✓ Assign one or more reviewer(s).
   ✓ Link to a Jira issue, and/or other GitHub issues or PRs.
   ✓ In the PR description delete any empty sections
     and all text commented in <!--, so that this text does not appear
     in merge commit messages.

* Don't waste reviewers' time:
   ✓ If it's a draft, select the Create Draft PR option.
   ✓ Self-review your changes to make sure nothing unexpected slipped through.

* Try to make your intent clear:
   ✓ Write a good Description that explains what this PR is meant to do.
   ✓ Jira will detect and link to this PR once created, but you can also
     link this PR in the description of the corresponding Jira ticket.
   ✓ Highlight what Testing you have done.
   ✓ Acknowledge any changes required to the Documentation.
-->

adp-1190

- [x] I have laxed condition when checking if tx is partially signed. We take into consideration only unique inputs now
- [x] Integration test

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

paweljakubas committed:
select witnesses only for unique inps

Commit:
TRANS_NEW_JOIN_02 integration test

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2415,7 +2415,7 @@ submitTransaction ctx apiw@(ApiT wid) apitx@(ApiSerialisedTransaction (ApiT seal
     let ourInps = getOurInps apiDecoded
 
-    let allInpsNum = length $ apiDecoded ^. #inputs
+    let allInpsNum = length $ L.nubBy sameInput $ apiDecoded ^. #inputs
```
```suggestion
    let allInpsNum = length $ L.nubBy samePaymentKey $ apiDecoded ^. #inputs
```

paweljakubas replied:
done

paweljakubas committed:
update how needed wintesses are count

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2477,6 +2481,20 @@ submitTransaction ctx apiw@(ApiT wid) apitx@(ApiSerialisedTransaction (ApiT seal
+    sameInput inp1 inp2 = case (inp1, inp2) of
+        (WalletInput (ApiWalletInput _ _ _ derPath1 _ _), WalletInput (ApiWalletInput _ _ _ derPath2 _ _) ) ->
+               derPath1 == derPath2
+        (ExternalInput txin1, ExternalInput txin2) ->
```
for external inputs we really cannot do better than that. we don't have knowledge of derivation paths. So in essence, we are not exact here. We provide reasonable upper bound. I will think about it when adding support for partial signing though

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2477,6 +2481,20 @@ submitTransaction ctx apiw@(ApiT wid) apitx@(ApiSerialisedTransaction (ApiT seal
+        WalletDelegationCertificate (QuitPool _) -> True
+        DelegationCertificate (JoinPoolExternal _ _) -> True
+        DelegationCertificate (QuitPoolExternal _) -> True
+        _ -> False
```
registration of stake credential does not need witness. Also I omitted witnesses from pool ops, mirs, etc.

paweljakubas committed:
renaming

piotr-iohk reviewed (approved):
Lgtm. Integration test passes, so does e2e tests that used to fail.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2415,11 +2415,15 @@ submitTransaction ctx apiw@(ApiT wid) apitx@(ApiSerialisedTransaction (ApiT seal
+    let inpsWits = length $ L.nubBy samePaymentKey $
+            (apiDecoded ^. #inputs) ++ (apiDecoded ^. #collateral)
+    let wdrlWits = length $ apiDecoded ^. #withdrawals
+    let certWits = length $ filter certWitNeeded $ apiDecoded ^. #certificates
```
This appears to ignore any potential for overlap of required witnesses between all of `inpsWits`, `wdrlWits` `certWits` (aside from among the inputs themselves).

So I believe this will overestimate the required number of witnesses if either:
1. A withdrawal key is identical to a input payment key
2. A certificate key is identical to a input payment key
3. There are multiple certs with the same key
4. There's a cert and withdrawal with the same key
5. Multiple withdrawals with the same key (absurd, not sure if technically possible)

I think (4) could be a common occurrence: users re-delegating and the wallet automatically withdrawing their rewards at the same time.

Anviking replied:
<s>Locally when I run `cabal v2-test cardano-wallet:integration --test-option=-j --test-option="10" --test-option="--match" --test-option "TRANS_NEW"` I do get a failure 
<img width="1376" alt="Skärmavbild 2022-02-08 kl  13 20 05" src="https://user-images.githubusercontent.com/304423/152985771-309da10b-2d06-44f3-83f0-68ed8179d797.png">

because the new tx workflow test for joining does [wait for rewards](https://github.com/input-output-hk/cardano-wallet/blob/baf862039942c05ab0b5e78874f66df51b742a2e/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs#L2258-L2282) before rejoining.

In CI it appears to have passed because of the retrying `it` 😰 </s>

Anviking replied:
Hm… never mind, I spoke too quickly. There are some Plutus related failures in CI however, and I guess the `retry` error logs are related to them.

Anviking replied:
Looking into why then [this waiting for rewards](https://github.com/input-output-hk/cardano-wallet/blob/baf862039942c05ab0b5e78874f66df51b742a2e/lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs#L2258-L2282) still works with this PR, plus trying to change the wallet type to a `rewardWallet` for extra rewards in the beginning, it seems that the `expectField #withdrawals (`shouldNotBe` [])` fails for the first two joins… I.e. there are no actual withdrawals 🤔 (I thought there would be)

paweljakubas replied:
it looks like there are needed 2 pools (as in hte test we are rejoining), and sometimes the second pool is not available

Anviking replied:
I pushed an integration test demonstrating (4)

```
 src/Test/Integration/Framework/DSL.hs:2446:7:
  1) API Specifications, NEW_SHELLEY_TRANSACTIONS, TRANS_NEW_JOIN_01e - Can re-join and withdraw at once
       expected a successful response but got an error: ClientError (Object (fromList [("code",String "missing_witnesses_in_transaction"),("message",String "The transaction has 3 inputs and 2 witnesses included. Submit fully-signed transaction.")]))
       While verifying value:
         ( Status
             { statusCode = 403
             , statusMessage = "Forbidden"
             }
         , Left
             ( ClientError
                 ( Object
                     ( fromList
                         [
                             ( "code"
                             , String "missing_witnesses_in_transaction"
                             )
                         ,
                             ( "message"
                             , String "The transaction has 3 inputs and 2 witnesses included. Submit fully-signed transaction."
                             )
                         ]
                     )
                 )
             )
         )
```
and a simple fix: let's not count certs and withdrawals towards the number of required witnesses for now.

Anviking committed:
Add test: TRANS_NEW_JOIN_01e - Can re-join and withdraw at once

Anviking committed:
Drop check of wdrl and cert witnesses

I believe the current logic will overestimate the required number of witnesses if either:

1. A withdrawal key is identical to a input payment key
2. A certificate key is identical to a input payment key
3. There are multiple certs with the same key
4. There's a cert and withdrawal with the same key
5. Multiple withdrawals with the same key (absurd, not sure if technically possible)

In particular (4) could be a common occurrence: users re-delegating and the wallet automatically
withdrawing their rewards at the same time. An integration test was
introduced with the previous commit.

This commit is the simplest fix for the test introduced in the previous
commit.

Anviking reviewed (approved):
lgtm after https://github.com/input-output-hk/cardano-wallet/pull/3110#discussion_r801639726

Anviking committed:
fix haddock error

paweljakubas committed:
even better calculation of needed number of wits
